dialog/form.go

```diff
@@ -0,0 +1,98 @@
+	"fyne.io/fyne/layout"
+	"fyne.io/fyne/theme"
+	"fyne.io/fyne/widget"
+)
+
+// FormDialog is like the standard Dialog but with confirm and cancel buttons, as well as the FormItems in the dialog.
+type FormDialog struct {
+	*dialog
+	items   []*widget.FormItem
+	confirm *widget.Button
+	cancel  *widget.Button
+}
+
+// validateItems acts as a validation edge state handler that will respond to an individual widget's validation
+// state before checking all others to determine the net validation state. If the error passed is not nil, then the
+// confirm button will be disabled. If the error parameter is nil, then all other Validatable widgets in items are
+// checked as well to determine whether the confirm button should be disabled.
+// This method is passed to each Validatable widget's SetOnValidationChanged method in items by NewFormDialog.
+func (d *FormDialog) validateItems(err error) {
+	if err != nil {
+		d.confirm.Disable()
+	} else {
+		for _, item := range d.items {
+			if validatable, ok := item.Widget.(fyne.Validatable); ok {
+				if err := validatable.Validate(); err != nil {
+					d.confirm.Disable()
+					return
+				}
+			}
+		}
+		d.confirm.Enable()
+	}
+}
+
+// NewFormDialog creates and returns a dialog over the specified application using
+// the provided FormItems. The cancel button will have the dismiss text set and the confirm button will
+// use the confirm text. The response callback is called on user action after validation passes.
+// If any Validatable widget reports that validation has failed, then the confirm
+// button will be disabled. The initial state of the confirm button will reflect the initial
+// validation state of the items added to the form dialog.
+// The MinSize() of the CanvasObject passed will be used to set the size of the window.
+func NewFormDialog(title, confirm, dismiss string, items []*widget.FormItem, callback func(bool),
+	parent fyne.Window) *FormDialog {
+	var itemObjects = make([]fyne.CanvasObject, len(items)*2)
+	for ii, item := range items {
+		itemObjects[ii*2] = widget.NewLabel(item.Text)
+		itemObjects[ii*2+1] = item.Widget
+	}
+	content := fyne.NewContainerWithLayout(layout.NewFormLayout(), itemObjects...)
+	d := &dialog{content: content, title: title, icon: nil, parent: parent}
+	d.callback = callback
+	// TODO: Copied from NewCustomConfirm above.
+	// This is still a problem because commenting out the `.Show()` call below will still result in the
+	// dialog being shown.
+	d.sendResponse = true
+
+	d.dismiss = &widget.Button{Text: dismiss, Icon: theme.CancelIcon(),
+		OnTapped: d.Hide,
+	}
+	confirmBtn := &widget.Button{Text: confirm, Icon: nil, Importance: widget.HighImportance,
+		OnTapped: func() {
+			d.hideWithResponse(true)
+		}}
+	// Mitigation for issue #1553
+	confirmBtn.SetIcon(theme.ConfirmIcon())
+	formDialog := &FormDialog{
+		dialog:  d,
+		items:   items,
+		confirm: confirmBtn,
+		cancel:  d.dismiss,
+	}
+	formDialog.validateItems(nil)
+
+	for _, item := range items {
```

Comment:
The operation within this loop could probably be moved into the one at the top of the function.

Reply:
That would be kind of tricky to put together since the items are being operated on by a struct that needs the items to exist before they can be operated upon, and the dialog also needs them to fill out its content, which is needed by the struct. Seems like a lot of sacrifice to readability to eliminate a one-time loop. How would you recommend this be done?

Reply:
Good point. Did not think about the contents withing that function that is set.

Reply:
Let's just keep it as is.